contrib/gdk-pixbuf/loader.c

```diff
@@ -187,11 +188,184 @@ static gboolean load_increment(gpointer data, const guchar * buf, guint size, GE
+    long quality = 52; /* default; must be between 0 and 100 */
+    gboolean save_alpha;
+    avifImage *avif;
+    avifRGBImage rgb;
+    avifResult res;
+    avifRWData raw = AVIF_DATA_EMPTY;
+    avifEncoder *encoder;
+    guint maxThreads;
+
+    if (f == NULL || pixbuf == NULL) {
+        return FALSE;
+    }
+
+    if (keys && *keys) {
+        gchar **kiter = keys;
+        gchar **viter = values;
+
+        while (*kiter) {
+            if (strcmp(*kiter, "quality") == 0) {
+                char *endptr = NULL;
+                quality = strtol(*viter, &endptr, 10);
+
+                if (endptr == *viter) {
+                    g_set_error(error,
+                                GDK_PIXBUF_ERROR,
+                                GDK_PIXBUF_ERROR_BAD_OPTION,
+                                "AVIF quality must be a value between 0 and 100; value “%s” could not be parsed.",
+                                *viter);
+
+                    return FALSE;
+                }
+
+                if (quality < 0 || quality > 100) {
+
+                    g_set_error(error,
+                                GDK_PIXBUF_ERROR,
+                                GDK_PIXBUF_ERROR_BAD_OPTION,
+                                "AVIF quality must be a value between 0 and 100; value “%d” is not allowed.",
+                                (int)quality);
+
+                    return FALSE;
+                }
+            } else {
+                g_warning("Unrecognized parameter (%s) passed to AVIF saver.", *kiter);
+            }
+
+            ++kiter;
+            ++viter;
+        }
+    }
+
+    if (gdk_pixbuf_get_bits_per_sample(pixbuf) != 8) {
+        g_set_error(error,
+                    GDK_PIXBUF_ERROR,
+                    GDK_PIXBUF_ERROR_UNKNOWN_TYPE,
+                    "Sorry, only 8bit images are supported by this AVIF saver");
+        return FALSE;
+    }
+
+    width = gdk_pixbuf_get_width(pixbuf);
+    height = gdk_pixbuf_get_height(pixbuf);
+
+    if ( width == 0 || height == 0) {
+        g_set_error(error,
+                    GDK_PIXBUF_ERROR,
+                    GDK_PIXBUF_ERROR_CORRUPT_IMAGE,
+                    "Empty image, nothing to save");
+        return FALSE;
+    }
+
+    save_alpha = gdk_pixbuf_get_has_alpha(pixbuf);
+
+    if (save_alpha) {
+        if ( gdk_pixbuf_get_n_channels(pixbuf) != 4) {
+            g_set_error(error,
+                        GDK_PIXBUF_ERROR,
+                        GDK_PIXBUF_ERROR_UNKNOWN_TYPE,
+                        "Unsupported number of channels");
+            return FALSE;
+        }
+    }
+    else {
+        if ( gdk_pixbuf_get_n_channels(pixbuf) != 3) {
+            g_set_error(error,
+                        GDK_PIXBUF_ERROR,
+                        GDK_PIXBUF_ERROR_UNKNOWN_TYPE,
+                        "Unsupported number of channels");
+            return FALSE;
+        }
+    }
+
+    max_quantizer = AVIF_QUANTIZER_WORST_QUALITY * ( 100 - CLAMP(quality, 0, 100)) / 100;
```

Comment:
I'm sure you're aware, but this strategy probably doesn't really produce a smooth quality dial (and neither does my semi-arbitrary mapping in colorist). The "right answer" is going to be a more nuanced combination of min+max, only I don't know it yet.

Someday I hope to come up with a smoother mapping to a single quality value, and I hope this doesn't set a weird precedent for people.

Reply:
Of course, when you will have better calculations, change it.
I wanted to avoid using complicated curves and chose linear functions only. And for very low quality I start to increase maxQuantizerAlpha, otherwise HEIC achieve smaller file.

I observed that quality depends on image size too. For small images, the minQuantizer is more important, for large images, maxQuantizer has higher impact. When there is big distance between minQ and maxQ, we have little control over compression. My distance between minQ and maxQ is hot higher than 20 (still grant some freedom to encoder but not too much).
Based on my experiments I think it is quite good, but I am open to changes.

Only some apps have write support and only some of them will set the quality parameter. Most of the time, the default values will be used anyway.